fix: NetworkAnimator rising RTT issue

NoelStephensUnity opened the pull request:
This fixes the issue with the rising RTT values discovered when testing the develop branch with BossRoom and now the `NetworkAnimator.CheckForAnimatorChanges` is no longer adding to the list but pulling from it, modifying, and then applying the modified AnimationState back to the same index position like the `NetworkAnimator.ServerSynchronizeNewPlayer` method was doing.  This includes a better approach to handling updating dirty states.

The testing for this is a bit tricky, but I did add some modifications to the trigger test where it makes sure the AnimationStates list within the AnimationMessage was not increasing in size over time (a part of the RTT bug was it would add as opposed to apply the state for each AnimationState change detected).

[MTT-4803](https://jira.unity3d.com/browse/MTT-4803)

## Testing and Documentation
- Includes integration test update.

NoelStephensUnity committed:
fix

This fixes the issue where NetworkAnimator was not completely updated to the changes with AnimationMessage.  NetworkAnimator will now only send states that have changed.

NoelStephensUnity committed:
test

adding additional checks to assure or AnimationStates list does not increase in size over time.

NoelStephensUnity committed:
Merge branch 'develop' into fix/networkanimator-rising-rtt

jeffreyrainy commented on com.unity.netcode.gameobjects/Components/NetworkAnimator.cs:
```diff
@@ -227,7 +227,7 @@ private void BuildDestinationToTransitionInfoTable()
         }
 
         /// <summary>
-        /// Creates the
```
ugh? That was missing like this??

NoelStephensUnity replied:
Yes, that week was a bit rough for me (family issue) and I made some really bad mistakes. 
😞

jeffreyrainy commented on com.unity.netcode.gameobjects/Components/NetworkAnimator.cs:
```diff
@@ -392,28 +391,15 @@ internal struct AnimationMessage : INetworkSerializable
-                }
-            }
+            // Determines how many AnimationState entries are dirty
+            // this is a 1:1 mapping of the AnimationStates indices
```
This comment is not as clear as it could be.
If it is a count, how can it also be a mapping what does it map to what?

NoelStephensUnity replied:
The List<AnimationState> AnimationStates indices values... 
Example:
If you have an IsDirtyCount of 5 then AnimationStates[0] through AnimationStates[4] would contain the states to be updated.  I will see if I can improve that comment.  When initializing the AnimationStates list and synchronizing a newly joined client, I use the layers count and wanted to make sure that it was clear that AnimationStates is not a 1:1 mapping of layers but is used to hold the AnimationState for each layer that had changed.

jeffreyrainy commented on com.unity.netcode.gameobjects/Components/NetworkAnimator.cs:
```diff
@@ -797,31 +789,32 @@ internal void CheckForAnimatorChanges()
-                };
+                // We use the IsDirtyCount here because we are only updating the AnimationStates that
+                // need to be updated
+                var animationState = m_AnimationMessage.AnimationStates[m_AnimationMessage.IsDirtyCount];
```
Echoes to my above comment. How is it that the count is always the one we want to index? It gives the impression that `AnimationStates` is actually a list of state to synchronize and, here, we mean to access the last? The the naming is misleading. `AnimationStates` conveys it being indexed by something logical around the animation itself.

Maybe I'm just unfamiliar, of maybe it needs a comment block at the top to explain what is indexed by what.

NoelStephensUnity replied:
Yeah, I might remove the comments... it possibly creates too much confusion.
IsDirtyCount is always reset to zero (i.e. starting with the 0 index of AnimationMessage.AnimationStates) prior to checking for changes in each layer's animation state and then when a layer's animation state change is detected it will fill out the AnimationState of the current IsDirtyCount index and then increment it.

NoelStephensUnity committed:
Merge branch 'develop' into fix/networkanimator-rising-rtt

NoelStephensUnity committed:
update

Removed the check for AnimationMessage.AnimationStates being null as it will always be null when deserializing.
Updated several comments for readability and clarity.

NoelStephensUnity committed:
style

noticed the local var was no longer a valid name.
renaming animMsg to animationState.

NoelStephensUnity committed:
Merge branch 'develop' into fix/networkanimator-rising-rtt

netcode-ci-service committed:
Merge develop into fix/networkanimator-rising-rtt
